Retain comments in field parser

leana8959 opened the pull request:
This is the first part of the exact print parser. In this PR I changed the lexer so instead of dropping the comments it emits them to the parser which is further stored in `GenericPackageDescription`.

Please let me know your thoughts!

---

Checklist below:


**This PR modifies [behaviour or interface](https://github.com/cabalism/cabal/blob/master/CONTRIBUTING.md#changelog)**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
  * [ ] [Is the change significant?](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#is-my-change-significant) If so, remember to add `significance: significant` in the changelog file.
* [x] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

andreabedini reviewed (commented):
Thank you for taking on this work. I think adding a `Comment` constructor to `Field` is not a good design. It modifies the meaning of the type (and indeed this forces you to make functions like `elementInLayoutContext` return `[Fields]`).

I think this has been already discussed before: we have the `ann` parameter which can be used to keep hold on the comments. E.g.
```haskell
data Comment ann = Comment !ann !ByteString
type FieldWithComments ann = Field ([Comment ann], ann)
```
In this design each Field carries the comments preceding it, annotated with their position. An extra annotation marks the position of the field itself. Any comment at the end of the file would need to be captured separately.

This is already the practice of few packages developed by the community. Is there a reason to deviate from this?

Bodigrim commented:
> That looks very interesting, but how would I deal with files that are just comments?

Are they valid Cabal files if there is nothing but comments? I don't think so.

> Could you elaborate which packages are these? I would love to have more insight on how people solve similar problems.

For instance, [`cabal-add`](https://hackage-content.haskell.org/package/cabal-add-0.2/docs/src/Distribution.Client.Rename.html#annotateFieldsWithSource) has a function 

```haskell
annotateFieldsWithSource :: ByteString -> [Field Position] -> [Field ByteString]
```

which annotates each field with its source including all adjacent comments. It can be modified to return `[Field (Position, ByteString)` if you need both.

Bodigrim commented on Cabal-syntax/src/Distribution/FieldGrammar.hs:
```diff
@@ -97,12 +100,26 @@ partitionFields = finalize . foldl' f (PS mempty mempty mempty)
     match _ = Nothing
+
+-- | Collect comments into a map. The second field of the output will have no comment
+extractComments :: Ord ann => [Field ann] -> (Map.Map ann ByteString, [Field ann])
```
Could you possibly outline how the output of this function is supposed to be used? Now that we detached comments from fields, how do we reconstruct the original document? How do we do it if `[Field ann]` is programmatically updated (say, adding or removing elements)?

leana8959 replied:
> Could you possibly outline how the output of this function is supposed to be used?

This function recursively extracts the comments from the fields (the updated design variant does the same thing, albeit being more concise).

It is used to extract the comments out right before running the `parseFieldGrammar` parser in `src/Distribution/PackageDescription/Parsec.hs`. This allows us to not change the grammar and be able to directly inject the comments into the `GenericPackageDescription`.

> Now that we detached comments from fields, how do we reconstruct the original document?

We can't yet. Following [Jappie's proposal](https://github.com/haskellfoundation/tech-proposals/blob/068a62315527c39f6bb58df58800060939e4542d/proposals/0000-cabal-exact-printer.md), we need to store the exact position of each field as a map (called `exactPositions`) indexed by the path in the rose tree (typed `[NameSpace]`, a list of path segments). And then we will have the right information to construct the exact printer.

> How do we do it if [Field ann] is programmatically updated (say, adding or removing elements)?

To quote Jappie:
> The issue for addition is that you now have to invent exact positions. for removal, if it involves a line, you've to fix up all following lines, (and it has to know something was removed).

I haven't thought about this thoroughly because it's out of scope of this PR, but it should be very much feasible.

jappeace commented:
Hi friends, thanks for all your responses. Leana needs some time to read up on the exact proposal to see how it all fits together before replying.
After chatting with her, I think she wants to go with Andrea's design for the comment field parser. As you can see, she's deeply in the weeds about many of the details of the parser; she even corrected some of the field grammar comments!
I don't know what you guys think, but I think it's good progress 🚀

leana8959 commented:
Thank you Bodigrim and Jappie for your kind words! That means a lot to me, I'm glad to be on the right track.

I have started (and completed) to rewrite my PR using Andrea's approach.
Right now the behaviour is identical to my old approach, while comments are tracked in the annotation `ann`.

> Are they valid Cabal files if there is nothing but comments? I don't think so.

Good to know. Currently the top level parser drops the comments consumed if there are no fields to attach them to.

Let me know what you think about the change :)

jappeace commented:
This PR isn't about that,
and we're intending to keep everything working. 🙂

phadej commented:
This PR does add

```haskell
  , exactComments :: ExactComments Position
```
  
  field to GPD. I don't see a point of having comments in GPD. (As noted in tests, it "breaks" equality)

ulysses4ever commented:
@mpickering can you take a high-level look at the design in this PR and tell us your opinion? The gist of it is:

```
-- Cabal-syntax/src/Distribution/Fields/Field.hs

data Comment ann = Comment !ByteString !ann
  deriving (Show, Generic, Eq, Ord, Functor)

data WithComments ann = WithComments
  { justComments :: ![Comment ann]
  , unComments :: !ann
  }
  deriving (Show, Generic, Eq, Ord, Functor)
```

and then many parsing utilities that used to return `... Field Position` now return `Field (WithComments Position)`. With casual renamings like below:

```diff
-parseGenericPackageDescription'
+parseAnnotatedGenericPackageDescription'
  :: Maybe CabalSpecVersion
  -> [LexWarning]
  -> Maybe Int
-  -> [Field Position]
-  -> ParseResult src GenericPackageDescription
-parseGenericPackageDescription' scannedVer lexWarnings utf8WarnPos fs = do
+  -> [Field (WithComments Position)]
+  -> ParseResult src AnnotatedGenericPackageDescription
+parseAnnotatedGenericPackageDescription' scannedVer lexWarnings utf8WarnPos fs = do
```

ulysses4ever commented:
One particular thing that bothers me is that clients will have to call `unComment` even if they don't care about comments (at least, it appears to be the case to me after looking at the diff in this patch). I feel like there should be a more graceful way to handle this. On the cabal meeting today @jappeace suggested that we could duplicate (I assume, some) parsing utilities so that some functions have two versions: one comment-bearing and another one, comment-less. I must say I quite like this idea. @geekosaur pointed out that this "just" moves the pressure from client's to our shoulders. But I feel like many clients aren't interested in comments anyway (they are surviving somehow today), and the cost of this sort of contained duplication (inside one module and very uniform) is low. So, overall, the cost/benefit ratio of this duplication-based idea is better than in the current proposal to me.

Any thoughts?

Meta-comment: I looked through the [tech proposal](https://github.com/haskellfoundation/tech-proposals/blob/main/proposals/0000-cabal-exact-printer.md) again, and there doesn't appear to be a technical description of a solution for this particular comments issue. It is totally fine, because the proposal would turn into a foliant at that level of detail. Neveretheless, I wish that, before doing all the technical work in this PR, the authors had discussed the actual technical solution for the particular issue (like preserving comments; others are listed in the tech proposal) on the Cabal bug tracker (here).

leana8959 commented:
I cleaned up everything and made sure the test passes, this is ready for review!

leana8959 commented:
A checklist on what has been discussed in this thread and their conclusions based on the current design and implementation. 

- [x] Andrea: Don't add a new Comment constructor
    -> use `ann` type variable to store comments

- [x] Bodigrim: Outline the usage of `exactComments`
    -> for now it extracts the comments annotation away from `[Field (WithComments Position)]`, leaving `[Field Position]`.
    The extracted `Comments` are made available further down the pipeline, and will be used by the following PRs.

- [x] phadej: Putting comments in GPD will break extensionality of `Eq` instance.
    -> we separate the comments and they are not stored in the GPD.

- [x] ulysses4ever: One particular thing that bothers me is that clients will have to call unComment
  even if they don't care about comments (at least, it appears to be the case to me after looking at
  the diff in this patch).
    -> jappeace: suggested that we could duplicate (I assume, some) parsing utilities so that some
    functions have two versions: one comment-bearing and another one, comment-less.

    -> leana: it would be possible to define the more complex one that has comments normally. Then, the one that without comments is defined as the commented one with the comments stripped away. These functions should not be complex to maintain, and saves downstream from managing comments.

philderbeast commented on Cabal-syntax/Cabal-syntax.cabal:
```diff
@@ -18,6 +18,11 @@ build-type:     Simple
 extra-doc-files:
   README.md ChangeLog.md
 
+flag CABAL_PARSEC_DEBUG
```
`CABAL_PARSEC_DEBUG` has been there already for 10 years so it is good to surface this as a flag rather than use another name:

https://github.com/haskell/cabal/blob/e7aaad898284c85fe566dd7edc738bb37c1e77c2/Cabal-syntax/src/Distribution/Fields/Lexer.x#L13

https://github.com/haskell/cabal/blob/e7aaad898284c85fe566dd7edc738bb37c1e77c2/Cabal-syntax/src/Distribution/Fields/LexerMonad.hs#L140-L142

philderbeast replied:
That said, all our existing flags are lower case:

```
$ grep -E "^(f|F)lag" **/*.cabal
cabal-install-solver/cabal-install-solver.cabal:flag debug-expensive-assertions
cabal-install-solver/cabal-install-solver.cabal:flag debug-tracetree
cabal-install/cabal-install.cabal:Flag native-dns
cabal-install/cabal-install.cabal:Flag lukko
cabal-install/cabal-install.cabal:flag git-rev
cabal-install/cabal-install.cabal:flag legacy-comparison
...
Cabal/Cabal.cabal:flag git-rev
```

leana8959 replied:
Renamed the flag to kebab case, just like the other ones :)

ulysses4ever commented on Cabal-syntax/src/Distribution/Fields/Lexer.x:
```diff
@@ -230,7 +230,9 @@ lexToken = do
         t <- action pos len_bytes inp
-        --traceShow t $ return tok
+#ifdef CABAL_PARSEC_DEBUG
+        traceShow t $ return tok
```
this will fail with the change removing the import of Debug.Trace above?

https://github.com/haskell/cabal/pull/11252/changes#diff-b8b1a6854a8674ea1c3da715560bdad742a6dee893b8c8dd0279d31a069118a8L34

leana8959 replied:
Cabal has its own `traceShow` in `Distribution.Compat.Prelude` which is why it compiles just fine.

ulysses4ever commented on Cabal-syntax/src/Distribution/Fields/Field.hs:
```diff
@@ -44,11 +51,26 @@ import qualified Data.Foldable1 as F1
+
+mapComments :: ([Comment ann] -> [Comment ann]) -> WithComments ann -> WithComments ann
+mapComments f (WithComments cs x) = WithComments (f cs) x
+
```
wouldn't it be wiser to have a callback of type `Comment ann -> Comment ann` instead? That would hide the list implementation detail.

```suggestion
mapComments :: (Comment ann -> Comment ann) -> WithComments ann -> WithComments ann
mapComments f (WithComments cs x) = WithComments (map f cs) x
```

leana8959 replied:
Sometimes with the parser branching the comments are parsed separately, and I have to insert them back in on the resulting node. If I use the signature you proposed I won't be able to do `++` with this function.

ulysses4ever commented on Cabal-syntax/src/Distribution/Fields/Field.hs:
```diff
@@ -44,11 +51,26 @@ import qualified Data.Foldable1 as F1
   = Field !(Name ann) [FieldLine ann]
   | Section !(Name ann) [SectionArg ann] [Field ann]
-  deriving (Eq, Show, Functor, Foldable, Traversable)
+  deriving (Eq, Show, Functor, Foldable, Traversable, Generic)
```
are these Generic instances really necessary? a quick reminder: we pay compilation time for every instance, and for Generic, these are sometimes nontrivial...

leana8959 replied:
This is because I have golden tests of `Field ann`, and by default `ToExpr` uses `Generic`. Do you think it'd be better to move Generic into `Data.TreeDiff.Instances.Cabal`? 

https://github.com/haskell/cabal/pull/11252/changes#diff-0c6d66260048e75eb3fccf82690f9283641c82c1ec27b4641e74dc8e1add91edR94-R95

ulysses4ever replied:
I suppose it's alright as is for now. Just wanted to make sure it's used.

ulysses4ever commented on Cabal-syntax/src/Distribution/Fields/Parser.hs:
```diff
@@ -166,46 +175,52 @@ indentOfAtLeast (IndentLevel i) = try $ do
 newtype LexerMode = LexerMode Int
 
+-- | This would change the state of the lexer and make interpretations of tokens different!
+-- Certain lexer states are unreachable without it.
```
I don't understand this comment. It looks like it needs (much) more context.

leana8959 replied:
Indeed, I updated it.

ulysses4ever commented on Cabal-syntax/src/Distribution/Fields/Parser.hs:
```diff
@@ -166,46 +175,52 @@ indentOfAtLeast (IndentLevel i) = try $ do
 newtype LexerMode = LexerMode Int
 
+-- | This would change the state of the lexer and make interpretations of tokens different!
+-- Certain lexer states are unreachable without it.
 inLexerMode :: LexerMode -> Parser p -> Parser p
 inLexerMode (LexerMode mode) p =
   do setLexerMode mode; x <- p; setLexerMode in_section; return x
 
 -----------------------
 -- Cabal file grammar
 --
+-- The non-terminals of the following grammar (symbols starting in uppercase)
+-- have their corresponding parser of the same name, starting with lowercase
+-- letter.
 
 -- $grammar
 --
 -- @
--- CabalStyleFile ::= SecElems
+-- CabalStyleFile             ::= Elements
 --
--- SecElems       ::= SecElem* '\\n'?
--- SecElem        ::= '\\n' SecElemLayout | SecElemBraces
--- SecElemLayout  ::= FieldLayout | FieldBraces | SectionLayout | SectionBraces
--- SecElemBraces  ::= FieldInline | FieldBraces |                 SectionBraces
--- FieldLayout    ::= name ':' line? ('\\n' line)*
--- FieldBraces    ::= name ':' '\\n'? '{' content '}'
--- FieldInline    ::= name ':' content
--- SectionLayout  ::= name arg* SecElems
--- SectionBraces  ::= name arg* '\\n'? '{' SecElems '}'
+-- Elements                   ::= Elements* '\\n'?
+-- Element                    ::= '\\n' ElementInLayoutContext
+--                              | ElementInNonLayoutContext
+-- ElementInLayoutContext     ::= FieldLayout | FieldBraces | SectionLayout | SectionBraces
+-- ElementInNonLayoutContext  ::= FieldInline | FieldBraces |                 SectionBraces
+-- FieldLayout                ::= name ':' line? ('\\n' line)*
+-- FieldBraces                ::= name ':' '\\n'? '{' content* '}'
+-- FieldInline                ::= name ':' content
+-- SectionLayout              ::= name arg* Elements
+-- SectionBraces              ::= name arg* '\\n'? '{' Elements '}'
```
Is this simply a renaming? I do like the new names better, but I sincerely hope that someone checked the diff to the symbol...

leana8959 replied:
It's not only a renaming, I also added where comments could appear in this grammar. I renamed the non-terminals because they were named differently from the parser, making understanding the codebase a lot harder than it needs to be.

I made sure this is correct as best as I could.

leana8959 replied:
I suggest using [patdiff](https://github.com/janestreet/patdiff) which highlights precisely which word changed and ignore the whitespace changes.


(Friendly note that this image might be out of date at some point and you should run patdiff or a similar tool on your own.) 

<img width="652" height="426" alt="image" src="https://github.com/user-attachments/assets/5f57a58f-79bc-434f-9b45-b0b5f503162f" />

ulysses4ever commented on Cabal-syntax/src/Distribution/PackageDescription/Parsec.hs:
```diff
@@ -72,12 +74,28 @@ import qualified Text.Parsec as P
+--
+-- TODO: re-benchmark, whether `deepseq` is important (both cabal-benchmarks and solver-benchmarks)
+-- TODO: remove the need for deepseq if `deepseq` in fact matters
+-- NOTE: IIRC it does affect (maximal) memory usage, which causes less GC pressure
```
When are you planning to address these: before or after the merge? The parser performance is of great importance (I am told)

leana8959 replied:
This comments was there before this PR, I had to move it around because I changed where the `deepseq` call is placed.

ulysses4ever commented:
>> Bodigrim: Outline the usage of exactComments
>
> -> for now it extracts the comments annotation away from [Field (WithComments Position)], leaving [Field Position].
> The extracted Comments are made available further down the pipeline, and will be used by the following PRs.

(NB: I assume Bodigrim meant `exTRactComments`)

I'm sorry to say but this outline is not helpful to me. Can you come up with concrete examples of what separated comments may be useful for? Because to me (like Bodigrim, I assume) they seem useless.

Same goes for `parseCommentedGenericPackageDescription`, I assume.

It'd be good to have some indication how this work connects to the more general exact-print work (concretely, #11690 or #11814, whichever is alive; btw it'd be good to garbage collect at some point).

leana8959 commented:
> Can you come up with concrete examples of what separated comments may be useful for? Because to me (like Bodigrim, I assume) they seem useless.

@ulysses4ever thanks a lot for the review!
Right now we expose the comment-preserving parsing functions because they might be useful for people who want to test it out. I could mark these functions as "unstable", what do you think?

As a different approach, we can also hide this from the interface for now. This PR will still have its value because it makes the parser ready for future exact print PRs even though this feature is not yet exposed via the interface.

I plan to feed the `Fields (WithComments Position)` to `ParsecFieldGrammar` and populate each field with their own associated comments.

```diff
 data ParsecFieldGrammar s a = ParsecFG
   { fieldGrammarKnownFields :: !(Set FieldName)
   , fieldGrammarKnownPrefixes :: !(Set FieldName)
-  , fieldGrammarParser :: forall src. (CabalSpecVersion -> Fields Position -> ParseResult src a)
+  , fieldGrammarParser :: forall src. (CabalSpecVersion -> Fields (WithComments Position) -> ParseResult src a)
   }
   deriving (Functor)
```
[src](https://github.com/haskell/cabal/blob/07b96a2578ae173a6919c8b29016cbc652b9b036/Cabal-syntax/src/Distribution/FieldGrammar/Parsec.hs#L115-L120)


This will allow printing each field in field grammar to be self contained, and easier to reason with. We also avoid the difficulty of correlating comments with modified fields, which was one of the big blockers in the trivia tree approach https://github.com/haskell/cabal/pull/11425.

tdammers commented:
I'm probably way late to the party, but I'm wondering whether it is morally correct to attach comments to fields in the first place.

AFAIK, the cabal file format does not explicitly specify that comments are supposed to be associated with whatever non-comment follows; by convention, most "real" comments in practice are, but when, for example, you comment out fields or sections in a cabal file, this is almost certainly not the case (you don't think of a commmented-out field as being associated with whatever field happens to come after it). The fact that comments can occur at the end of a cabal file and need to be handled specially there is, IMO, a symptom of this, and my gut feeling says that this probably means that associating comments with fields like this is not the right design.

leana8959 commented:
@tdammers 

> you comment out fields or sections in a cabal file, this is almost certainly not the case (you don't think of a commmented-out field as being associated with whatever field happens to come after it).

Why not? Note that currently cabal doesn't retain any comments. It is very "correct" because it avoids answering the question, but it also doesn't let us print out the comments.


Making comments a proper construction on par with Field and Sections was the previous design. However, as andrea bedini pointed out, it [modifies the meaning of the type](https://github.com/haskell/cabal/pull/11252#pullrequestreview-3338602055). 

Also, distinguishing what comment is a real comment is not in the scope of this PR. We don't do haddock-like generation where we need to know to which AST node the comment is attached to. This PR is only concerned with storing the comments properly and passing them on to be printed back out completely.

> The fact that comments can occur at the end of a cabal file and need to be handled specially there is, IMO, a symptom of this, and my gut feeling says that this probably means that associating comments with fields like this is not the right design.

(Just to make sure we are on the same page, can you point to me which part of the diff you are talking about?)

I think this situation occurs in every language that allows you to put a comment before _and_ after a line. The real problem is that when the cabal file is only comments, we'd have no fields to attach them to. This is discussed [here](https://github.com/haskell/cabal/pull/11252#issuecomment-3413266386) and we agreed that we'd only support valid cabal files, which will have a non null amount of fields/sections of fields.

leana8959 commented:
I removed the exported "comment extraction" function as we decided in the cabal weekly meeting, and fixed the warnings from hlint. Let me know when it's all good and I'll rebase :)

Mikolaj commented:
Might this be a good moment for a (re-)review?

jappeace commented:
@Mikolaj yes! I wanted to bring this up again in the cabal meeting.

leana8959 committed:
Retain comments in field parser

This lays the foundation for Cabal-exactprint to consume user comments.

The parser of the envelope format ("field parser") now annotates each of
the parsed fields with its preceding and succeeding comments, along with
the existing source location annotation 'Position'.

This change is orthogonal to the final chosen exactprint implementation,
or the algebra we use to modify existing fields and/or generate new
fields. Instead, it merely makes the user comments available to any
consumer that reads cabal fields, the envelope format.

readFields* functions now have their counterparts that parses with
comments, named readFieldsWithComments*.

Bodigrim commented on Cabal-syntax/src/Distribution/Fields/Field.hs:
```diff
@@ -42,11 +51,26 @@ import qualified Data.Foldable1 as F1
 -- Cabal file
 -------------------------------------------------------------------------------
 
+data Comment ann = Comment !ByteString !ann
```
What are the typical instantiations of `ann` across Cabal's code base? Could they be described in a comment?

leana8959 replied:
`ann` currently is either left uninstantiated or instantiated as `Position`. It makes sense for `Comment` to have a `Position`, but in retrospective I can also encode `Position` directly (and remove `ann`) inside. What do you think?

Bodigrim replied:
If it's the same as `Field ann` is typically instantiated then perhaps you can leave `Comment ann` as is.

Bodigrim commented on Cabal-syntax/src/Distribution/Fields/Field.hs:
```diff
@@ -42,11 +51,26 @@ import qualified Data.Foldable1 as F1
+data Comment ann = Comment !ByteString !ann
+  deriving (Show, Generic, Eq, Ord, Functor)
+
+data WithComments ann = WithComments
```
This can use a comment describing the meaning of the data type and its fields.

Bodigrim commented on Cabal-syntax/src/Distribution/Fields/Field.hs:
```diff
@@ -42,11 +51,26 @@ import qualified Data.Foldable1 as F1
+mapComments f (WithComments cs x) = WithComments (f cs) x
+
+mapCommentedData :: (ann -> ann) -> WithComments ann -> WithComments ann
+mapCommentedData f (WithComments cs x) = WithComments cs (f x)
```
* I'd expect the names of these function to be related to the names of record fields. 
* Perhaps defining them as `Lens' (WithComments ann) _` (from `Distribution.Compat.Lens`) would be a better choice, because `Lens` imply that there is exactly one element to focus on.

leana8959 replied:
Good call, I have changed it.

Bodigrim commented on Cabal-syntax/Cabal-syntax.cabal:
```diff
@@ -18,6 +18,11 @@ build-type:     Simple
   README.md ChangeLog.md
 
+flag cabal-parsec-debug
+  description: Enable debug build for the cabal field lexer/parser.
```
Let's not add such debug flags, they are publicly visible, and for a package like `Cabal` are likely to cause more confusion and harm than save us time during development.

leana8959 replied:
This flag was already there, and the supporting code was already existing in cabal. Should I still remove it?

Bodigrim replied:
Sorry, I don't follow. The `master` branch does not seem to contain such flag?..

leana8959 replied:
My bad, I wasn't very clear. I meant that in the Haskell code there was already existing CPP code for tracing guarded behind the macro `CABAL_PARSEC_DEBUG`. What I did was adding a cabal flag that exposed this CPP macro and fixing the compilation.

Bodigrim replied:
CPP can stay, a daring soul can pass `-DCABAL_PARSEC_DEBUG`. I'm opposed only to exposing it even further by providing a public flag.

leana8959 replied:
Sounds reasonable to me, done :)

Bodigrim commented on Cabal-syntax/src/Distribution/Parsec/Position.hs:
```diff
@@ -16,10 +19,11 @@ data Position
 
 instance Binary Position
 instance NFData Position
+instance Structured Position
```
What requires this instance now? Same question about `instance Data Position` above.

(There is no `instance Structured Fields` as far as I can see, so I'm confused why `instance Structured Position` is needed. And pretty much nothing uses `class Data`.)

leana8959 replied:
Thanks, nice catch. I removed the instance. I had a test that needed this but removed it.

Bodigrim commented on Cabal-syntax/src/Distribution/Fields/Parser.hs:
```diff
@@ -223,26 +250,89 @@ inLexerMode (LexerMode mode) p =
-  return es
+  case es of
+    Left' _ -> pure [] -- We discard the comments here, because it is not a valid cabal file
+    Right' es' -> pure es'
```
It does not seem to me that the mental overhead of our own `Either'` is justified: it is used only few times, which seemingly can be replaced with bang patterns.

leana8959 replied:
In retrospective I think it might be a premature optimization, I removed `Either'` directly.

Bodigrim commented on Cabal-syntax/src/Distribution/Fields/Parser.hs:
```diff
@@ -223,26 +250,89 @@ inLexerMode (LexerMode mode) p =
+commentsAfter p = do
+  x <- p
+  postCmts <- many tokComment
+  pure $ fmap (WithComments postCmts) x
```
Could it be `commentsAfter p = flip WithComments <$> p <*> many tokComment`?

The real question though is: if comments in `WithComments` are the ones following and not preceding, should not they go to the second field instead of the first one?

leana8959 replied:
> Could it be `commentsAfter p = flip WithComments <$> p <*> many tokComment?`

~~Indeed, nice catch.~~

It would need to be `flip (fmap . WithComments) <$> p <*> many tokComment` but I feel like it's less readable.

> The real question though is: if comments in WithComments are the ones following and not preceding, should not they go to the second field instead of the first one

It is not that easy, because when we parse something with trailing comments, we don't yet know if there are fields that follow. For example, when parsing some field lines, we don't know if the comments we got are for this fieldline or if there are following fields and the comments belong there.

To implement what you meant (on which I agree is the "better" outcome), I'd need to guess that there's a following field, and then backtrack to attaching to current field line if I guessed wrong. Personally I think that obfuscates the flow of the parser and may cause obscure bugs that only manifest when user writes comments a certain way. Considering we don't anchor comments to AST nodes like haddock does and comments attachments to nodes aren't specified at all [^1], I think this is reasonable.

In case in the future we need to attach comments to the later field (should it exist), I also think it's a better idea to implement that with a function that post-processes the fields, just to keep the parser simple.


[^1]:
    It is possible to give comments that have ambiguous attachments like the following. 
    ```cabal
    build-depends:
      foo
      
      -- some comments
    exposed-modules:
      Main
    ```

Bodigrim replied:
Perhaps I was not clear what I meant. When we have `WithComments "foo" "bar"`, does it stand for `foo -- bar` or for `--bar \n foo`?

leana8959 replied:
Interesting question. The current semantic is "around", so we can't tell without inspecting the `Position` annotation. Do you think it'd be useful to make `WithComments` to have two fields, one for leading comments and one for trailing comments?

Bodigrim replied:
I think I'm missing something fundamental here. If the current semantics is "around" then what is stored where? Is it the case that both preceding and following comments will be elements of the list `justComments :: ![Comment ann]`? Could there be less than two or more than two elements? Is the order of elements of the list predetermined or can one put them in any order, even if not matching locations?

I think having `priorComment :: Comment ann` and `followingComment :: Comment ann` would be handy and self-explanatory, if it is achievable.

leana8959 replied:
In cabal the lexer emits a `Comment` token for a line comment. Also, all comments are line comments. This is why the type is `[Comment ann]` instead of `Comment ann`.

I also think it might make manipulating the comments a bit easier during printing, I'll do that. Thanks for the suggestion! I'll keep you updated.

leana8959 replied:
@Bodigrim I tried to split comments representation into two fields. Though, due to how the parse tree is structured, only the first field will have leading comments. The rest all have trailing comments. This might be a bit weird for the users if we distinguish them while one of two kinds is almost always empty. What do you think?

leana8959 committed:
document types and their methods

leana8959 committed:
rewrite mapComments and mapCommentData using lens

leana8959 committed:
remove unused instances

leana8959 committed:
remove strict either

leana8959 committed:
run forumolu

Bodigrim commented on Cabal-syntax/src/Distribution/Fields/Parser.hs:
```diff
@@ -164,46 +172,64 @@ indentOfAtLeast (IndentLevel i) = try $ do
 
+-- | Cabal's lexer is generated using Alex. It has multiple different states ("start-codes").
+-- Each state has its own rules and interprets the input stream differently.
+-- This has the benefit of providing more exact error message by having more
```
I'd imagine that first of all it has a benefit of lexing the input stream correctly ;)

Bodigrim commented on Cabal-syntax/src/Distribution/Fields/Parser.hs:
```diff
@@ -164,46 +172,64 @@ indentOfAtLeast (IndentLevel i) = try $ do
+-- This has the benefit of providing more exact error message by having more
+-- knowledge about what symbol should follow in the input stream.
+--
+-- This lexer never switches to some of the states itself.
```
To which states the lexer can never switch itself? On a brief inspection I didn't find any.

leana8959 replied:
I admit that this is not easy to spot. However I hesitate to add the specific states into the comments because I'm afraid that it might get out of sync. In [this blog post](https://blog.haskell.org/a-comment-preserving-cabal-parser/#tracing-out-lexer-states) I wrote about which state are not reachable. TL;DR: switching to `in_field_layout` and `in_field_braces` from `in_section` is only done in the parser.

Do you think I should detail this in the comments?

Bodigrim replied:
Looking at `Distribution.Fields.Lexer`, I see both `setStartCode in_field_layout` and `setStartCode in_field_braces`, so it can switch to these states on its own. Am I missing something?

leana8959 replied:
Correct, `bol_field_layout` and `in_field_layout` can switch to each other, but without something that switches from `in_section` to `in_field_layout` the lexer would reach neither `bol_field_layout` nor `in_field_layout` states.

Bodigrim replied:
OK, then I'd suggest to say exactly that, because the current wording is confusing. All states are reachable from the lexer itself, it's just that certain transitions of state are not.

leana8959 replied:
I can surely reword it, but

> All states are reachable from the lexer itself, it's just that certain transitions of state are not.

is incorrect.

Here's an illustration stolen from my [comment-preserving cabal parser article](https://blog.haskell.org/a-comment-preserving-cabal-parser/). The blue transitions are done by the field parser, while the red transitions are done by the lexer. You can see that without the field parser, you don't reach two whole classes of states, namely `{in,bol}_field_layout` and `{in,bol}_field_braces`.
<img width="1339" height="304" alt="image" src="https://github.com/user-attachments/assets/f813e2ee-1562-4468-913d-be2b295236d7" />

Bodigrim replied:
Thanks for the correction. I suggest to expand the comment (my confusion is a sure sign that other readers might have the same misunderstanding), maybe even with some ASCII art, but up to you.

leana8959 replied:
Indeed, I think adding a ASCII flow chart in the comments would help people :)

Bodigrim commented on Cabal-syntax/src/Distribution/Fields/Parser.hs:
```diff
@@ -223,26 +249,89 @@ inLexerMode (LexerMode mode) p =
+
+-- | Collect in annotation one or more comments after a parser succeeds
+-- Careful with the 'Functor' instance!
+-- If you use this with Field you might attach the same comments everywhere
```
Sounds like a good reason not to generalize to any `Functor` then?.. What are the instantiations of `f`?

leana8959 replied:
You have a point! It's only used for `FieldLine`, so I rewrote it.

leana8959 committed:
remove use of Functor in commentsAfter

leana8959 committed:
don't expose cabal-parsec-debug flag
